README.md
# brotli plugin for webpack

[![Greenkeeper badge](https://badges.greenkeeper.io/mynameiswhm/brotli-webpack-plugin.svg)](https://greenkeeper.io/)

This plugin compresses assets with [Brotli](https://github.com/google/brotli) compression algorithm using [zlib](https://nodejs.org/api/zlib.html#zlib_zlib_brotlicompress_buffer_options_callback), [iltorb](https://github.com/MayhemYDG/iltorb#brotliencodeparams) or [brotli.js](https://github.com/foliojs/brotli.js) libraries for serving it with [ngx_brotli](https://github.com/google/ngx_brotli) or such.

## Installation

```
npm install --save-dev brotli-webpack-plugin
```

## Usage

``` javascript
var BrotliPlugin = require('brotli-webpack-plugin');
module.exports = {
	plugins: [
		new BrotliPlugin({
			asset: '[path].br[query]',
			test: /\.(js|css|html|svg)$/,
			threshold: 10240,
			minRatio: 0.8
		})
	]
}
```

Arguments:

* `asset`: The target asset name. Defaults to `'[path].br[query]'`.
  * `[file]` is replaced with the original asset file name.
  * `[fileWithoutExt]` is replaced with the file name minus its extension, e.g. the `style` of `style.css`.
  * `[ext]` is replaced with the file name extension, e.g. the `css` of `style.css`.
  * `[path]` is replaced with the path of the original asset.
  * `[query]` is replaced with the query.
* `test`: All assets matching this RegExp are processed. Defaults to every asset.
* `threshold`: Only assets bigger than this size (in bytes) are processed. Defaults to `0`.
* `minRatio`: Only assets that compress better that this ratio are processed. Defaults to `0.8`.
* `deleteOriginalAssets`: remove original files that were compressed with brotli. Default: false

Optional arguments for Brotli (see [iltorb](https://github.com/MayhemYDG/iltorb#brotliencodeparams) doc for details):
* `mode`: Default: 0,
* `quality`: Default: 11,
* `lgwin`: Default: 22,
* `lgblock`: Default: 0,
* `size_hint`: Default: 0,
* `disable_literal_context_modeling`: Default: false

## License

Heavily copy-pasted from [webpack/compression-webpack-plugin](https://github.com/webpack/compression-webpack-plugin) by [Tobias Koppers](https://github.com/sokra).

Licensed under [MIT](./LICENSE).


CHANGELOG.md
# Change Log

### 1.1.1
Add safe compilation hoks to prevent deprecation warning for frozen compilation assets in Webpack 5, along with backwards compatibility for older Webpack versions.

### 1.1.0
Added native brotli support via [zlib.brotliCompressSync](https://nodejs.org/api/zlib.html#zlib_zlib_brotlicompress_buffer_options_callback) for Node v11.7+.

### 1.0.0
Added native support for Webpack 4 hooks API (to avoid `DeprecationWarning`), along with backwards compatibility for legacy Webpack versions.

### 0.5.0
Added `deleteOriginalAssets` option to remove original files that were compressed with brotli.

### 0.4.3
Updated `iltorb`s  available configuration parameters in documentation.

### 0.4.2
Updated [iltorb](https://github.com/MayhemYDG/iltorb) to 2.0.1.

### 0.4.1
Added warning message when `iltorb` is not available.

### 0.4.0
Moved compilation from `this-compilation` to `emit` stage. See [issue #179](https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/179) of [extract-text-webpack-plugin](https://github.com/webpack-contrib/extract-text-webpack) for details.

### 0.3.0
Optional [brotli.js](https://github.com/devongovett/brotli.js) fallback when [iltorb](https://github.com/MayhemYDG/iltorb) is not available/couldn't be compiled.

### 0.2.0
Added support for `[fileWithoutExt]` and `[ext]` placeholders. This allows a mapping from e.g. 'style.css' to 'style.br.css', aiding use with tools that infer Content-Type from extension.

### 0.1.0
Initial release


test/index.test.js
var BrotliPlugin = require('../index.js');
var webpack = require('webpack');
var iltorb = require('iltorb');
var rmRf = require('rimraf');

var OUTPUT_DIR = __dirname + '/tmp/';

function createCompiler(options) {
    var defaultOptions = {
        bail: true,
        cache: false,
        entry: __dirname + '/fixtures/entry.js',
        output: {
            path: OUTPUT_DIR,
            filename: '[name].js',
            chunkFilename: '[id].[name].js'
        },
        plugins: [
            new BrotliPlugin()
        ]
    };

    if (!Array.isArray(options)) {
        options = Object.assign({}, defaultOptions, options);
    }

    return webpack(options);
}

function compile(compiler) {
    return new Promise(function (resolve, reject) {
        compiler.run(function (err, stats) {
            if (err) return reject(err);
            resolve(stats);
        });
    });
}

describe('when applied with default settings', function () {
    beforeEach(function (done) {
        rmRf(OUTPUT_DIR, done)
    });

    afterEach(function (done) {
        rmRf(OUTPUT_DIR, done)
    });

    it('compresses and decopresses', function () {
        var compiler = createCompiler();

        new BrotliPlugin().apply(compiler);

        return compile(compiler).then(function (stats) {
            expect(stats.compilation.assets).toHaveProperty(['main.js']);
            expect(stats.compilation.assets).toHaveProperty(['main.js.br']);

            var source = stats.compilation.assets['main.js'].source();
            expect(source).toContain('console.log');

            var decopressed = iltorb.decompressSync(stats.compilation.assets['main.js.br'].source()).toString();
            expect(decopressed).toMatch(source);
        });
    });
});


test/fixtures/entry.js
// foo
// bar
var a = 2 + 2;

module.exports = function Foo() {
    var b = 2 + 2;
    console.log(b + 1 + 2);
};

compress.js
function adapter() {
    try {
        var zlib = require('zlib');
        if (zlib.hasOwnProperty('brotliCompress')) {
            return zlib.brotliCompress;
        }
    } catch (err) {}

    try {
        console.warn('warning: couldn\'t find native brotli support in zlib library. trying to fall back to iltorb.');
        var iltorb = require('iltorb');
        return iltorb.compress;
    } catch (err) {
        console.warn('warning: couldn\'t load iltorb library. trying to fall back to brotli.js.');
        console.warn(err);

        try {
            var brotli = require('brotli');
            return function (content, options, callback) {
                var result = brotli.compress(content, options);
                callback(null, result);
            }
        } catch (err) {
            throw new Error('iltorb or brotli not found. See https://github.com/mynameiswhm/brotli-webpack-plugin for details.');
        }
    }
}

module.exports = adapter;


index.js
var async = require('async');
var url = require('url');

var RawSource = require('webpack-sources/lib/RawSource');

function BrotliPlugin(options) {
    options = options || {};
    this.asset = options.asset || '[path].br[query]';
    this.test = options.test || options.regExp;
    this.threshold = options.threshold || 0;
    this.minRatio = options.minRatio || 0.8;
    this.deleteOriginalAssets = options.deleteOriginalAssets || false;

    var compress = require('./compress.js')();

    var brotliOptions = {
        mode: options.hasOwnProperty('mode') ? options.mode : 0,
        quality: options.hasOwnProperty('quality') ? options.quality : 11,
        lgwin: options.hasOwnProperty('lgwin') ? options.lgwin : 22,
        lgblock: options.hasOwnProperty('lgblock') ? options.lgblock : 0,
        enable_dictionary: options.hasOwnProperty('enable_dictionary') ? options.enable_dictionary : true,
        enable_transforms: options.hasOwnProperty('enable_transforms') ? options.enable_transforms : false,
        greedy_block_split: options.hasOwnProperty('greedy_block_split') ? options.greedy_block_split : false,
        enable_context_modeling: options.hasOwnProperty('enable_context_modeling') ? options.enable_context_modeling : false
    };

    this.compress = function (content, callback) {
        compress(content, brotliOptions, callback);
    };
}
module.exports = BrotliPlugin;

BrotliPlugin.prototype.apply = function (compiler) {
    var emit = function (compilation, assets, callback) {
        async.forEach(Object.keys(assets), function (file, callback) {
            if (this.test && !this.test.test(file)) {
                return callback();
            }

            var asset = assets[file];
            var content = asset.source();
            if (!Buffer.isBuffer(content)) {
                content = Buffer.from(content, 'utf-8');
            }

            var originalSize = content.length;
            if (originalSize < this.threshold) {
                return callback();
            }

            this.compress(content, function (err, result) {
                if (err) {
                    return callback(err);
                }

                if (result.length / originalSize > this.minRatio) {
                    return callback();
                }

                var parse = url.parse(file);
                var sub = {
                    file: file,
                    fileWithoutExt: file.split('.').slice(0, -1).join('.'),
                    ext: file.split('.').slice(-1).join(''),
                    path: parse.pathname,
                    query: parse.query || ''
                };

                var newFile = this.asset.replace(/\[(file|fileWithoutExt|ext|path|query)]/g, function (p0, p1) {
                    return sub[p1];
                });
                if (this.deleteOriginalAssets) {
                    delete assets[file];
                }
                assets[newFile] = new RawSource(result);
                callback();
            }.bind(this));
        }.bind(this), callback);
    }.bind(this);

    if (compiler.hooks) {
    	if (compiler.hooks.thisCompilation) {
            const plugin = {
                name: 'BrotliPlugin',
                stage: compiler.webpack.Compilation.PROCESS_ASSETS_STAGE_OPTIMIZE_TRANSFER,
                additionalAssets: true
            };
            compiler.hooks.thisCompilation.tap(plugin.name, (compilation) => {
                compilation.hooks.processAssets.tapAsync(
                    plugin,
                    (assets, callback) => emit(compilation, assets, callback)
                );
            });
        } else {
            const plugin = { name: 'BrotliPlugin' };
            compiler.hooks.emit.tapAsync(
                plugin,
                (compilation, callback) => emit(compilation, compilation.assets, callback)
            );
        }
    } else {
        compiler.plugin('emit', emit);
    }
};
